### PR TITLE
Improvement

### PR DESCRIPTION
**Pull request by banasiakadam60**
1) Added creating an Alpha file
2) Some improvements

### CODE REVIEW

File: taskcollector/TaskCollector.cpp [hunk 1]
```diff
@@ -1,81 +1,78 @@
...
 		// no signature ?
 		// try to guess the file format from the file extension
 		fif = FreeImage_GetFIFFromFilename(lpszPathName);
 	}
 	// check that the plugin has reading capabilities ...
-	if((fif != FIF_UNKNOWN) && FreeImage_FIFSupportsReading(fif)) {
+	if ((fif != FIF_UNKNOWN) && FreeImage_FIFSupportsReading(fif)) {
 		// ok, let's load the file
-		FIBITMAP *dib = FreeImage_Load(fif, lpszPathName, flag);
+		auto dib = std::make_shared<FIBITMAP>(*FreeImage_Load(fif, lpszPathName, flag));
```
Comment (chfast) [inline taskcollector/TaskCollector.cpp]: I understand what you wanted to get here, but it was done incorrectly.

Shared pointer will just free the memory by `delete` operator. There is no guarantee that the memory was allocated in compatible way. You should always put resource back to libraries they come from.  FreeImage library expects the function `FreeImage_Unload()` to be called on `FIBITMAP*`.

To make it correct you can use create RAII wrapper for `FIBITMAP*` or use `std::unique_ptr` with a custom deleter that calls `FreeImage_Unload()`.

File: taskcollector/TaskCollector.cpp [hunk 2]
```diff
@@ -1,81 +1,78 @@
...
-	@param lpszPathName Pointer to the full file name
-	@param flag Optional save flag constant
-	@return Returns true if successful, returns false otherwise
+@param dib Pointer to the dib to be saved
+@param lpszPathName Pointer to the full file name
+@param flag Optional save flag constant
+@return Returns true if successful, returns false otherwise
 */
-bool GenericWriter(FIBITMAP* dib, const char* lpszPathName, int flag) {
+bool GenericWriter(bitmap_t dib, const char* lpszPathName, int flag) {
```
Comment (chfast) [inline taskcollector/TaskCollector.cpp]: There is no need to pass a copy of shared pointer to the function. The function need not to be the owner of the bitmap.

File: taskcollector/TaskCollector.cpp [hunk 3]
```diff
@@ -1,81 +1,78 @@
...
 /** Generic image loader
-	@param lpszPathName Pointer to the full file name
-	@param flag Optional load flag constant
-	@return Returns the loaded dib if successful, returns NULL otherwise
+@param lpszPathName Pointer to the full file name
+@param flag Optional load flag constant
+@return Returns the loaded dib if successful, returns NULL otherwise
 */
-FIBITMAP* GenericLoader(const char* lpszPathName, int flag) {
+bitmap_t GenericLoader(const char* lpszPathName, int flag) {
```
Comment (chfast) [inline taskcollector/TaskCollector.cpp]: I would keep information that it is still a pointer. How about naming it `bitmap_ptr`?

File: taskcollector/TaskCollector.cpp [hunk 4]
```diff
@@ -85,44 +82,51 @@ void FreeImageErrorHandler(FREE_IMAGE_FORMAT fif, const char *message) {
...
 		return true;
 	};
 
 	bool addAlphaFile(const char* pathName, int flag = 0) {
-		FIBITMAP *img = GenericLoader(pathName, flag);
+		auto img = GenericLoader(pathName, flag);
 		if (img == NULL)
 			return false;
-		alphaChunks.push(img);
+		alphaChunks.push_back(img);
```
Comment (chfast) [inline taskcollector/TaskCollector.cpp]: This line crashes on Windows. The issue might be related to the usage of shared pointer I mentioned earlier.

Comment (badb) [inline taskcollector/TaskCollector.cpp]: Both images are generated properly on windows, but application throws an error and stops. 
From the debugger: 

> Unhandled exception at 0x77D16054 (ntdll.dll) in taskcollector.exe: 0xC0000374: Sterta została uszkodzona (parameters: 0x77D32378).

File: taskcollector/TaskCollector.cpp (hunk 3 above)
Comment (banasiakadam60) [inline taskcollector/TaskCollector.cpp]: OK, changed.

File: taskcollector/TaskCollector.cpp (hunk 1 above)
Comment (banasiakadam60) [inline taskcollector/TaskCollector.cpp]: You, you're right. I've added FreeImage_Unload() to the TaskCollector destructor, but I forgot to add this here.
shared_ptr changed to unique_ptr

File: taskcollector/TaskCollector.cpp (hunk 2 above)
Comment (banasiakadam60) [inline taskcollector/TaskCollector.cpp]: Changed

File: taskcollector/TaskCollector.cpp (hunk 4 above)
Comment (banasiakadam60) [inline taskcollector/TaskCollector.cpp]: I was checking my changes on debug build and I didn't notice this failure. This time I've checked both: debug and release build and nothing crashes on my PC

File: taskcollector/TaskCollector.cpp [hunk 5]
```diff
@@ -85,256 +97,239 @@ void FreeImageErrorHandler(FREE_IMAGE_FORMAT fif, const char *message) {
...
+	std::vector<bitmap_ptr> alphaChunks;
 
 public:
-	
-	bool addImgFile(const char* pathName, int flag = 0)  {
-		FIBITMAP *img = GenericLoader(pathName, flag);
-		if (img == NULL) 
+	TaskCollector() = default;
+	TaskCollector(const TaskCollector&) = delete;
+	TaskCollector(TaskCollector &&other)
```
Comment (chfast) [inline taskcollector/TaskCollector.cpp:108]: You could use default move constructor here `TaskCollector(TaskCollector&&) = default`. However, I'm not sure MSVC support that yet.

**Comment by chfast** [comment]
Now it looks great! And we don't have the crash any more!

**Pull request merged**